classes/send_message_form.php

```diff
@@ -0,0 +1,272 @@
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Tool messenger block admin form.
+ *
+ * @package block_evasys_sync
+ * @copyright 2017 Robin Tschudi
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_messenger;
+
+use moodleform;
+use html_writer;
+
+defined('MOODLE_INTERNAL') || die;
+
+require_once($CFG->libdir . '/formslib.php');
+
+class send_message_form extends moodleform {
+    protected function definition() {
+        $mform = $this->_form;
+
+        $mform->addElement('html', '<div id="adminsettings">');
+
+        $followuprequest = isset($_POST["followup"]) and !isset($_POST["sendmessagebutton"]);
+        if (!$followuprequest) {
+            $mform->addElement('hidden', 'followup', null);
+            $mform->setType('followup', PARAM_INT);
+            $mform->setDefault('followup', 0);
+        } else {
+            $name = 'followup';
+            $title = get_string('followupto', 'tool_messenger');
+            $mform->addElement('text', $name, $title);
+            $mform->setType($name, PARAM_INT);
+            $mform->addHelpButton($name, $name, 'tool_messenger');
+            $mform->setDefault($name, $_POST["followup"]);
+            $parent = new message_persistent($_POST["followup"]);
+        }
+        $mform->addElement('hidden', 'abort', null);
+        $mform->setType('abort', PARAM_INT);
+        $mform->setType('followupto', PARAM_INT);
+
+        $name = 'subject';
+        $title = get_string('subject', 'tool_messenger');
+        $mform->addElement('textarea', $name, $title, 'wrap="virtual" rows="1" cols="100"');
+        $mform->setType($name, PARAM_TEXT);
+
+        $name = 'message';
+        $title = get_string('message', 'tool_messenger');
+        $mform->addElement('editor', $name, $title,
+            array('enable_filemanagement' => false, 'maxfiles' => 0, 'noclean' => true, 'trusttext' => true));
+        $mform->setType($name, PARAM_RAW);
+
+        $name = 'recipients';
+        $title = get_string('sendto', 'tool_messenger');
+        $roles = $this->get_roles();
+        $mform->addElement('select', $name, $title, $roles)->setMultiple(true);
+        if ($followuprequest) {
+            $mform->disabledIf($name, 'followup', 'neq', 0);
+        }
+        $name = 'knockout_enable';
+        $title = get_string('knockoutenable', 'tool_messenger');
+        $mform->addElement('advcheckbox', $name, $title, ' ');
+        $default = 1;
+        if ($followuprequest) {
+            $default = $parent->get('knockoutdate') == 0;
```

Comment:
this is caused by moodle trying to hold data entered in forms until they are submitted, because of this the default value is updated,however the supplied value from the last iteration of the form takes priority over this. It should work in the newest version.